twill-core/src/main/java/org/apache/twill/internal/kafka/EmbeddedKafkaServer.java

```
@@ -29,20 +34,71 @@
  */
 public final class EmbeddedKafkaServer extends AbstractIdleService {
 
-  private final KafkaServerStartable server;
+  public static final String START_TIMEOUT_RETRIES = "twill.kafka.start.timeout.retries";
+
+  private static final Logger LOG = LoggerFactory.getLogger(EmbeddedKafkaServer.class);
+  private static final String DEFAULT_START_TIMEOUT_RETRIES = "5";
+
+  private final int startTimeoutRetries;
+  private final KafkaConfig kafkaConfig;
+  private KafkaServer server;
 
   public EmbeddedKafkaServer(Properties properties) {
-    server = new KafkaServerStartable(new KafkaConfig(properties));
+    this.startTimeoutRetries = Integer.parseInt(properties.getProperty(START_TIMEOUT_RETRIES,
+                                                                       DEFAULT_START_TIMEOUT_RETRIES));
+    this.kafkaConfig = new KafkaConfig(properties);
   }
 
   @Override
   protected void startUp() throws Exception {
-    server.startup();
+    int tries = 0;
+    do {
+      KafkaServer kafkaServer = createKafkaServer(kafkaConfig);
+      try {
+        kafkaServer.startup();
+        server = kafkaServer;
+      } catch (Exception e) {
+        kafkaServer.shutdown();
+        kafkaServer.awaitShutdown();
+
+        Throwable rootCause = Throwables.getRootCause(e);
+        if (rootCause instanceof ZkTimeoutException) {
+          // Potentially caused by race condition bug described in TWILL-139.
+          LOG.warn("Timeout when connecting to ZooKeeper from KafkaServer. Attempt number {}.", tries, rootCause);
+        }
```

Should we propagate any exceptions that are _not_ ZkTimeoutExceptions?  Or should we continue to retry those too?  Seems like we should let those out since we don't really know how to handle them.